launcher/src/main/java/org/apache/spark/launcher/CommandBuilder.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Internal interface that defines a command builder.
+ */
+interface CommandBuilder {
```

Comment:
IMO - it might be simpler to just not have this interface. It's confusing because a lot of the heavy lifting for building commands in in SparkLauncher, but that class doesn't actually implement this. Also, you never interact with `CommandBuilder`'s in a generic way, you only use it in a case where you already have a branch distinguishing the only two possible implementations.

In cases such as this, I'm not sure it the code simpler to understand when there is an interface.